pynumaflow/function/server.py

```diff
@@ -0,0 +1,92 @@
+import asyncio
+import logging
+from os import environ
+
+from pynumaflow.function import udfunction_pb2, Datum
+from pynumaflow.function import udfunction_pb2_grpc
+
+from google.protobuf import timestamp_pb2 as _timestamp_pb2
+
+import grpc
+from typing import Callable, Any
+
+from pynumaflow._constants import (
+    FUNCTION_SOCK_PATH,
+    DATUM_KEY,
+)
+from pynumaflow.function import Messages
+
+if environ.get("PYTHONDEBUG"):
+    logging.basicConfig(level=logging.DEBUG)
+
+_LOGGER = logging.getLogger(__name__)
+
+UDFMapCallable = Callable[[str, Datum, Any], Messages]
+
+
+class UserDefinedFunctionServicer(udfunction_pb2_grpc.UserDefinedFunctionServicer):
+    def __init__(self, map_handler: UDFMapCallable, sock_path=FUNCTION_SOCK_PATH):
+        self.__map_handler: UDFMapCallable = map_handler
+        self.sock_path = sock_path
+        self._cleanup_coroutines = []
+
+    def MapFn(self, request: udfunction_pb2.Datum, context):
+        """Applies a function to each datum element."""
+        key = ""
+        for metadata_key, metadata_value in context.invocation_metadata():
+            if metadata_key == DATUM_KEY:
+                key = metadata_value
+
+        msgs = self.__map_handler(
+            key,
+            Datum(
+                value=request.value,
+                event_time=request.event_time.event_time.ToDatetime(),
+                watermark=request.watermark.watermark.ToDatetime(),
+            ),
+        )
+
+        datum_list = []
+        for msg in msgs.items():
+            datum_list.append(udfunction_pb2.Datum(key=msg.key, value=msg.value))
+
+        return udfunction_pb2.DatumList(elements=datum_list)
+
+    def ReduceFn(self, request_iterator, context):
+        """Applies a reduce function to a datum stream."""
+        context.set_code(grpc.StatusCode.UNIMPLEMENTED)
+        context.set_details("Method not implemented!")
+        raise NotImplementedError("Method not implemented!")
+
+    def IsReady(self, request, context):
```

Comment:
```suggestion
    def is_ready(self, request, context):
```
Add typing please. Both args and return types